org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/IndexManager.java

```diff
@@ -1736,106 +1745,61 @@ public Optional<MetaIndex> getMetaIndex() {
 	}
 }
 
-void scheduleForMetaIndexUpdate(Index index) {
-	synchronized(this.metaIndexUpdates){
-		if (this.metaIndexUpdates.contains(index)) {
-			if (VERBOSE) {
-				Util.verbose("-> already waiting for meta-index update for " + index); //$NON-NLS-1$
-			}
-			return;
-		}
-		this.metaIndexUpdates.add(index);
-	}
-	requestIfNotWaiting(new MetaIndexUpdateRequest());
-}
-
-class MetaIndexUpdateRequest implements IJob {
-	volatile boolean isCancelled;
-
-	@Override
-	public boolean belongsTo(String jobFamily) {
-		return false;
-	}
-
-	@Override
-	public void cancel() {
-		this.isCancelled = true;
+private void writeIntoMetaIndex() {
+	if (DISABLE_META_INDEX) {
+		return;
 	}
 
-	@Override
-	public void ensureReadyToRun() {
-		// no op
-	}
+	Index index = null;
+	int metaIndexUpdatesSize;
+	boolean hasMoreIndexes = true;
 
-	@Override
-	public boolean execute(IProgressMonitor progress) {
-		while((progress == null || !progress.isCanceled()) && !this.isCancelled) {
-			Index index = null;
-			int metaIndexUpdatesSize;
-			synchronized(IndexManager.this.metaIndexUpdates){
-				Iterator<Index> iterator = IndexManager.this.metaIndexUpdates.iterator();
-				if (iterator.hasNext()) {
-					index = iterator.next();
-					iterator.remove();
-				}
-				metaIndexUpdatesSize = IndexManager.this.metaIndexUpdates.size();
-			}
-			if (index == null) {
-				return true;
-			}
-			if (index.monitor == null) {
-				// index got deleted since acquired
-				continue;
-			}
-			File indexFile = index.getIndexFile();
-			if (indexFile == null) {
-				continue;
-			}
-			if (VERBOSE) {
-				Util.verbose("-> meta-index update from queue with size " + metaIndexUpdatesSize); //$NON-NLS-1$
-			}
-			try {
-				updateMetaIndex(indexFile.getName(), index.getMetaIndexQualifications());
-			} catch (IOException e) {
-				Throwable cause = e.getCause();
-				if (cause != null) {
-					Util.log(e, "Failed to update meta index"); //$NON-NLS-1$
-				} else {
-					if (JobManager.VERBOSE) {
-						Util.verbose("-> failed to update meta index for index " + indexFile.getName() //$NON-NLS-1$
-								+ " because of the following exception:"); //$NON-NLS-1$
-						e.printStackTrace();
-					}
-				}
-			}
+	while (hasMoreIndexes) {
+		// the metaIndexUpdates is already a synchronized collection, so no need to synchronized following operations
```

Comment:
you still need to synchronize because hasNext() may be already wrong  when next() is called on iterator. Even though each Operation on the synchronizedSet is synchronized performing multiple operations is not atomic.

Reply:
Do you see there is a possibility for that to happen, because we are right now in the IndexLoop which is the thread which execute indexing. All search operation are also blocked until this method is executed.

Reply:
It is hard to see in the sourcecode which method is called within which thread, but i guess that metaIndexUpdates in accessed in multiple threads. for example during classpathchange which could happen while indexer is running:
![image](https://github.com/eclipse-jdt/eclipse.jdt.core/assets/51790620/f828fe6c-ffaf-4913-ae8d-5fd0bd7d1a6f)